### ISSUE TITLE
Support different arrow end points

### ISSUE THREAD

**Issue by owenconti** [excalidraw/excalidraw#616]
It would be awesome for relationship diagrams if arrows could support configurable arrow heads, so users could add relationship information to an arrow via "crow's foot" notation:

https://en.wikipedia.org/wiki/Entity–relationship_model#Crow.27s_foot_notation

My initial instinct earlier today was to be able to draw an arrow and then configure the arrow heads on each end of the arrow. Something like this (assume all arrows were drawn from left to right):

![Screen Shot 2020-01-29 at 5 28 00 PM](https://user-images.githubusercontent.com/791222/73409521-c8e5a700-42bc-11ea-857e-8826a41007b9.png)

Personally, I don't think there's much need in supporting the "one" and "zero" relations right away as I assume most people usually use the "many" notation.

If there's interest in supporting this feature, I'd be willing to look into implementing it.

**Comment by dwelle** [comment]
I like the idea of merging line & arrow together (https://github.com/excalidraw/excalidraw/issues/344#issuecomment-574737967), and allowing to switch end-point. Then this could be implemented without adding new shapes. The line/arrow should use the same implementation, anyway.

**Comment by vjeux** [comment]
We can merge the underlying implementation but let’s not merge the line and arrow icons. They serve two distinct purposes when drawing diagrams.

**Comment by dwelle** [comment]
Copying the issue description from #839 for completeness' sake

> Would be nice to have arrow start/end feature (or some variation of configuring an arrow) so that you don't end up with arrows like this:
> 
> !\[image\](https://user-images.githubusercontent.com/5770711/75201737-47582c00-5737-11ea-9a7a-a8003fe168b0.png)
> 
> This is how it looks in sketch:
> 
> !\[image\](https://user-images.githubusercontent.com/5770711/75202055-1f1cfd00-5738-11ea-85ef-3bacf418be18.png)
> 
> !\[image\](https://user-images.githubusercontent.com/5770711/75202063-23491a80-5738-11ea-9060-ffbaffa7da81.png)

**Comment by henno** [comment]
What is the current best workaround to get the double sided arrow?

**Comment by dai-shi** [comment]
![image](https://user-images.githubusercontent.com/490574/93711740-afff6e00-fb8b-11ea-826a-5eb925337656.png)

https://excalidraw.com/#json=5733360995401728,narze4RKoz1UmuXwbmhLyg

**Comment by carlsverre** [comment]
In all my work with excalidraw having double ended arrows is the only thing I have really missed so far.  Major :+1: from me

**Comment by steveruizok** [comment]
Mind if I implement this? [POC here](https://excalidraw-pi.vercel.app/) where all arrows are double-sided.

![Kapture 2020-12-02 at 10 27 43](https://user-images.githubusercontent.com/23072548/100860945-1278c280-3489-11eb-9adb-cec37c746654.gif)

The existing `getArrowPoints` function by @GasimGasimzada works just as well for reversed arrows: we just need to change which operations and points we're looking at. [See delta here](https://github.com/excalidraw/excalidraw/compare/master...steveruizok:arrow-options#diff-b5487856e823a67aae9b278f6ab4d7a588125d44bc539a0bcb1a8045b11bb549).

I can't find anything at breaks it, so if we wanted to move forward it would just be a matter of:
- deciding how to represent arrow head options in the data
- deciding on the UI

My guess would be that we would eventually remove the `"arrow"` element type and instead add "decorators" options to the linear element's head, tail, and possibly joins. As in other design tools, a user might give turn any line into an arrow (or vis versa) by changing its decorators. With that done, selecting the arrow tool would just be a quicker way to create a line with a pre-configured head decorator.

In the short term—and given that arrows also have other unique powers, such as shape binding—I could see adding these decorator options to the arrow element. The panel could then show a field to allow a user to change the selected arrow's start and end decorator. We could begin with either an arrow or null, or else I could also add a circle / square / etc as in other tools.

**Comment by dwelle** [comment]
Yea, this issue wasn't blocked on implem complexity, but entirely on the UI representation and UI complexity.

We've discussed this multiple times, across different PRs (dunno where) and chat. We couldn't gain consensus on whether it should be part of the properties menu, or be rendered as an extra menu below the toolbar (see below) when you select an arrow (a tool or an existing shape), or whether we should introduce it at all (I think we should).

![image](https://user-images.githubusercontent.com/5153846/100868293-b8352d00-349b-11eb-9e53-d3c2e2eccbe5.png)

Also we haven't agreed on whether to only support different directions (incl double-sided), or different arrow endings (caps) as well.

> deciding how to represent arrow head options in the data

It should be a prop on `ExcalidrawLinearElement` type, but whether we should encode both direction and endings into a single prop is the question. Since we'll likely start with direction first, and it's unclear whether we'll support customizing endings as well (see above), I'd probably go with separate props:

```
{
  arrowSide: /* default */ "end" | "start" | "both" | "none"
}
```

Not sure on the nomenclature:

- arrow direction — arrowSide/arrowType/arrowDirection/...
- arrow ending – arrowCap/arrowEnding/...

**Comment by lipis** [comment]
What about as a first implementation we support only one kind of arrow and adding one more section in the property and when clicking on start, end it will scroll through available endings (arrow and none in our first case)

<img width="256" alt="Screenshot 2020-12-02 at 5 27 45 PM" src="https://user-images.githubusercontent.com/125676/100893127-b2ebd880-34c3-11eb-8d34-3988464b1f93.png">

**Comment by steveruizok** [comment]
![Kapture 2020-12-03 at 10 50 35](https://user-images.githubusercontent.com/23072548/101000177-6cdf5500-3555-11eb-9d4d-50d4b046980c.gif)

I like it! [Try it here](https://excalidraw.steveruizok.vercel.app/).

There were some design choices in the code around how to represent arrow decorators. I've made my choices but I'd expect a few rounds of feedback and revisions once I get a PR up. Let me clean up the icons and labels and such and I'll get that done later this week.

**Comment by dai-shi** [comment]
@dwelle Can we consider again to use this UI for toggle binding?

**Comment by dwelle** [comment]
Um, to me I'd have no idea what to do, so it would be a trial and error process. And to me those active states don't make sense since either no-end-cap and arrow-end-cap are active states to me.

More importantly, since now those buttons encompass both direction and endings, if we later decide to add more types of endings you'd have to cycle through N arrow caps which isn't a good UX.

**Comment by lipis** [comment]
> More importantly, since now those buttons encompass both direction and endings, if we later decide to add more types of endings you'd have to cycle through N arrow caps which isn't a good UX.

You are right about that, but if we have 2-3 it should be fine and if we ever get more then we can think on it when time comes!

@steveruizok The demo is great! Could you create a PR already so we could comment directly there?

**Comment by steveruizok** [comment]
First step shipped! https://github.com/excalidraw/excalidraw/pull/2452

**Comment by dwelle** [comment]
@steveruizok personally I'd support all those you used in the PR initially — all useful.

**Comment by TylerLMcIntosh** [comment]
Did double-ended arrows ever get implemented?

**Comment by ahmadao2214** [comment]
@TylerLMcIntosh I literally came here to ask the same thing ahaha!!

**Comment by meshy** [comment]
It looks like yes! 
![image](https://github.com/user-attachments/assets/78af7329-2c55-4354-a9eb-936169ae8b8e)

**Comment by TylerLMcIntosh** [comment]
Amazing! Thanks to all who build that out, and thanks for figuring out where it hides in the interface @meshy

**Comment by krzemian** [comment]
A small UX improvement request on my side: it is currently unclear the two buttons work for arrowhead setting. I literally had to google it and only then had I realized, after coming here and toggling back and forth between github and the whiteboard that the little straight line is indeed a setting for the other arrowhead.

One example of how this could be solved would be to display a greyed out arrow head in the other icon, in place of a black straight line. This would achieve two objectives
- Inform that clicking this button leads to something related to arrowheads
- Indicate that it's currently inactive (grey)

Currently, the buttons focus on how the whole line looks (a straight line) rather than on focusing on the end points (pointy arrows)

**Comment by tmowad** [comment]
Thanks!  The diagram @meshy posted above was key, since...when looking at the UX it didn't occur to me that the [going left] and [going right] Arrowhead buttons could independently or both be turned on/off...
